crates/log_probe/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+use composer_api::{self, Client, LogStats};
+use log::{self, Level};
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        mpsc::{sync_channel, Receiver, SyncSender},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum LogProbeError {
+    #[error("Network error: {0}")]
+    NetworkError(String),
+}
+
+pub struct LogProbe {
+    tx: SyncSender<AggregatorMessage>,
+    shutdown: Arc<AtomicBool>,
+}
+
+impl Drop for LogProbe {
+    fn drop(&mut self) {
+        self.shutdown.store(true, Ordering::Relaxed);
+    }
+}
+
+enum AggregatorMessage {
+    AddRecord(Level),
+    Tick,
+}
+
+impl LogProbe {
+    pub fn new(
+        server_address: Option<String>,
+        report_frequency: Duration,
+    ) -> Result<Self, LogProbeError> {
+        let shutdown = Arc::<AtomicBool>::default();
+        let client = if let Some(address) = server_address {
+            Client::new(address)
+        } else {
+            Client::try_default()
+        }
+        .map_err(|e| LogProbeError::NetworkError(format!("{e}")))?;
+
+        let (tx, rx) = sync_channel::<AggregatorMessage>(100);
+        spawn_aggregator_thread(rx, client, shutdown.clone());
+        spawn_tick_thread(tx.clone(), shutdown.clone(), report_frequency);
+
+        Ok(Self { tx, shutdown })
+    }
+}
+
+fn spawn_tick_thread(
+    tx: SyncSender<AggregatorMessage>,
+    shutdown: Arc<AtomicBool>,
+    report_frequency: Duration,
+) {
+    std::thread::spawn(move || {
+        let start = Instant::now();
+        for deadline in (0..).map(|i| start + i * report_frequency) {
+            if let Err(e) = tx.send(AggregatorMessage::Tick) {
+                eprintln!("Failed to communicate with aggregator thread: {e}")
+            }
+
+            if shutdown.load(Ordering::Relaxed) {
+                break;
+            }
+
+            std::thread::sleep(deadline - Instant::now());
+        }
+    });
+}
+
+fn spawn_aggregator_thread(
+    rx: Receiver<AggregatorMessage>,
+    client: Client,
+    shutdown: Arc<AtomicBool>,
+) {
+    let mut log_stats = LogStats::default();
+    let mut report_start = Instant::now();
+
+    std::thread::spawn(move || {
+        for message in rx {
+            if shutdown.load(Ordering::Relaxed) {
+                break;
+            }
+
+            match message {
+                AggregatorMessage::AddRecord(record) => add_record(&mut log_stats, record),
+                AggregatorMessage::Tick => {
+                    log_stats.span = report_start.elapsed();
+                    if let Err(err) = client.send(&composer_api::Event::LogStats(log_stats)) {
+                        eprintln!("Could not send event {:?}", err)
+                    };
+                    log_stats = Default::default();
```

Comment:
Any reason not to use
```suggestion
                    log_stats = LogStats::default();
```
?

Reply:
There's a similar tradeoff to using `Self` vs the type name in things like constructors. using `Self` and `Default` just causes a smaller mess if you refactor the name :) but yeah, the price you pay is not seeing the type explicitly. 

In cases where the variable name is so close to the type name I usually do it like this, but I don't have a strong opinion (happy to change it).

Reply:
Right. Indeed the type was obvious to me - that's why the question even came to my mind. .)

Your heuristic makes total sense to me. Resolving. .)